test/functional/p2p_eviction.py

```diff
@@ -122,6 +129,178 @@ def run_test(self):
+        cur_mock_time += (HEADERS_RESPONSE_TIME + 10)
+        node.setmocktime(cur_mock_time)
+        self.log.info("Test that the peer gets evicted")
+        test_node.wait_for_disconnect()
+
+        self.log.info("Create an outbound connection and keep lagging behind, but not too much")
+        # Test that if the peer never catches up with our current tip, but it does with the
+        # expected work that we set when setting the timer (that is, our tip at the time)
+        # we do not disconnect the peer
+        test_node = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=0, connection_type="outbound-full-relay")
+        self.log.info("Mine a block so our peer starts lagging")
+        best_block_hash = self.generateblock(self.nodes[0], output="raw(42)", transactions=[])["hash"]
+        tip_header = from_hex(CBlockHeader(), node.getblockheader(best_block_hash, False))
+        test_node.sync_with_ping()
+        self.log.info("Keep catching up with the old tip and check that we are not evicted")
+        for _ in range(10):
+            # Advance time but not enough to evict the peer
+            best_block_hash = self.generateblock(self.nodes[0], output="raw(42)", transactions=[])["hash"]
+            cur_mock_time += (CHAIN_SYNC_TIMEOUT + 1)
+            node.setmocktime(cur_mock_time)
+
+            # Send a header with the old tip
+            headers_message = msg_headers()
+            headers_message.headers = [tip_header]
+            test_node.send_and_ping(headers_message)
+            tip_header = from_hex(CBlockHeader(), node.getblockheader(best_block_hash, False))
+
+            # Check that we are not evicted
+            cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
+            node.setmocktime(cur_mock_time)
+            test_node.sync_with_ping()
+
+        self.log.info("Create an outbound connection and take some time to catch up, but do it in time")
+        # Check that if the peer manages to catch up within time, the timeouts are removed (and the peer is not disconnected)
+        # We are reusing the peer from the previous case which already sent us a valid (but old) block and whose timer is ticking
+
+        # Send an updated headers message matching our tip
+        headers_message = msg_headers()
+        headers_message.headers = [from_hex(CBlockHeader(), node.getblockheader(node.getbestblockhash(), False))]
+        test_node.send_and_ping(headers_message)
+
+        # Wait for long enough for the timeouts to have triggered and check that we are still connected
+        cur_mock_time += (CHAIN_SYNC_TIMEOUT + 1)
+        node.setmocktime(cur_mock_time)
+        test_node.sync_with_ping()
+        cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
+        node.setmocktime(cur_mock_time)
+        self.log.info("Test that the peer does not get evicted")
+        test_node.sync_with_ping()
+
+        node.disconnect_p2ps()
+
+    def test_outbound_eviction_protected(self):
```

Comment:
Added notes below to provide insight into what was reviewed:

In `test_outbound_eviction_protected()` if unexpected eviction occurs (i.e. the P2P connection is no longer present), then `sync_with_ping()` will raise (`IOError('Not connected')`, causing the test to fail.  Moved node.disconnect_p2ps() above the last test_node.sync_with_ping() call and, as expected, this occurred.

nit: The heading comment in the `test_outbound_eviction_protected()` omits that a protected peer is also non-block-only, but this is implicit in the call to add_outbound_p2p_connection() using outbound-full-relay as the connection_type.

net_processing.cpp:
```
A peer is marked as protected if all of these are true:
 - its connection type is IsBlockOnlyConn() == false
 - it gave us a valid connecting header
 - we haven't reached MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT yet
 - its chain tip has at least as much work as ours
```

Reply:
Protected peer actually implies `outbound-full-relay`. The comment you mentioned may be outdated. I may add an update to it to the PR.

https://github.com/bitcoin/bitcoin/blob/edc14bc9da332cae39f8803db559b532b3c74e16/src/net_processing.cpp#L2875-L2885